### PR TITLE
Add Deep Dive RPC Chapter

### PR DESCRIPTION
**Pull request by Hoverbear**
Adds the RPC chapter from https://github.com/tikv/deep-dive-tikv/pull/30

### CODE REVIEW

**Comment by lucperkins** [comment]
Such cool content 😍

File: content/deep-dive/rpc/protobuf.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+KvPair {
+  key: "TiKV"
+  value: "Astronaut"
+}
+```
+
+When a message is actually *sent* between two applications, a binary format is
+used. You can learn more about the binary format in the [protobuf documentation](https://developers.google.com/protocol-buffers/docs/encoding).
+
+The message above would be an instance of a structure predefined in a `.proto`
```
Comment (dcalvin) [inline content/deep-dive/rpc/protobuf.md]: above, or below?
```suggestion
The message above is an instance of a structure predefined in a `.proto`
```

File: content/deep-dive/rpc/protobuf.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+If you have used [`serde_json`](https://docs.serde.rs/serde_json/) or another
+JSON library, you may have experienced the task of defining schemas for
+structures. This becomes a maintenance burden as your infrastructure grows to
+span many languages.
+
+You need to do this with protocol buffers as well, but you only do it once, and
+the protobuf compiler will generate bindings for any language it knows how to.
+
+Protobuf generates this code in a backwards compatible manner. If an application
```
Comment (dcalvin) [inline content/deep-dive/rpc/protobuf.md]: "this" is an unclear reference here. 
```suggestion
Protobuf generates code in a backwards compatible manner. If an application
```

File: content/deep-dive/rpc/protobuf.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+finds unfamiliar data is isn't familiar with, it just ignores them. This allows
```
Comment (dcalvin) [inline content/deep-dive/rpc/protobuf.md]:
```suggestion
finds data it isn't familiar with, it just ignores them. This allows
```

File: content/deep-dive/rpc/protobuf.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+for a safe evolution of an API.
+
+## More than just data
+
+Protobufs also enable the definition of *services*. This allows the
+definition of RPC calls in the `*.proto` files.
+
+This example demonstrates a service called `ScanService` which provides a remote
```
Comment (dcalvin) [inline content/deep-dive/rpc/protobuf.md]: Split the sentence to make it more readable. 
```suggestion
This example demonstrates a service called `ScanService`. It provides a remote
```

File: content/deep-dive/rpc/_index.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+standards.
+
+### Network Protocols
+
+The vast majority of services work over the HTTP or HTTP/2 network protocols.
+This solves problems such as:
+
+* [URIs]
+* [Sessions]
+* [Status Codes] (eg. 404 Not Found)
```
Comment (dcalvin) [inline content/deep-dive/rpc/_index.md]:
```suggestion
* [Status Codes] (e.g. 404 Not Found)
```

File: content/deep-dive/rpc/_index.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+* [Methods] (eg. GET, POST, PUT)
```
Comment (dcalvin) [inline content/deep-dive/rpc/_index.md]:
```suggestion
* [Methods] (e.g. GET, POST, PUT)
```

File: content/deep-dive/rpc/_index.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+* [Headers] (eg. Authorization, User-Agent)
```
Comment (dcalvin) [inline content/deep-dive/rpc/_index.md]:
```suggestion
* [Headers] (e.g. Authorization, User-Agent)
```

File: content/deep-dive/rpc/_index.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+With those abilities supported, there remains a need to work with structures of
+data. Commonly this ends up being an *interface description language* format
+like [Protocol Buffers] (protobufs), which TiKV uses. Unlike an *interchange*
+*format*, an *interface description language* allows for the definition of
+services and RPCs in addition to just data interchange. This solves the
+following problems:
+
+* [Encoding] (ASCII? UTF-8? UTF-16? WTF-8?)
+* Serialization/Deserialization format (Text-to-`struct` and vice versa)
+* Backward/Forward compatibility (e.g. Field data type changed)
```
Comment (ice1000) [inline content/deep-dive/rpc/_index.md]: Compatibility among different versions of the same `struct`

File: content/deep-dive/rpc/_index.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+With those abilities supported, there remains a need to work with structures of
+data. Commonly this ends up being an *interface description language* format
+like [Protocol Buffers] (protobufs), which TiKV uses. Unlike an *interchange*
+*format*, an *interface description language* allows for the definition of
+services and RPCs in addition to just data interchange. This solves the
+following problems:
+
+* [Encoding] (ASCII? UTF-8? UTF-16? WTF-8?)
+* Serialization/Deserialization format (Text-to-`struct` and vice versa)
```
Comment (ice1000) [inline content/deep-dive/rpc/_index.md]: To-structural object? Not sure, but not all programming languages call that `struct`.

File: content/deep-dive/rpc/grpc.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+service StreamScan {
+  rpc BatchGet (string, string) returns (stream string) {};
+}
+```
+
+## A Rust Crate
+
+One of the TiKV maintaining organizations, PingCAP, maintains the
+[`grpc-rs`](https://github.com/pingcap/grpc-rs) library, providing Rust bindings
+to the C library of gRPC. TiKV uses this crate.
```
Comment (ice1000) [inline content/deep-dive/rpc/grpc.md]: to the official gRPC implementation in C

File: content/deep-dive/rpc/protobuf.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
+---
+title: Protocol Buffers
+weight: 2
+---
+Interface definition languages, such as protobufs, are most commonly used to
+store and transmit data between applications.
+
+They define a way to *serialize* structures as text, as well as *de-serialize*
```
Comment (ice1000) [inline content/deep-dive/rpc/protobuf.md]: Why hyphen? I noticed you're using the word "deserialize" directly somewhere above.

File: content/deep-dive/rpc/grpc.md (hunk 10 above)
Comment (ice1000) [reply content/deep-dive/rpc/grpc.md]: Btw, we're actually using a modified version of the official implementation. Links might be helpful.

File: content/deep-dive/rpc/_index.md (hunk 9 above)
Comment (Hoverbear) [reply content/deep-dive/rpc/_index.md]: I think we're taking a fairly Rust centric perspective, and it's a generic enough term.

File: content/deep-dive/rpc/grpc.md (hunk 10 above)
Comment (Hoverbear) [reply content/deep-dive/rpc/grpc.md]: We already link to the repo which is 1 click away from the grpc submodule: https://github.com/pingcap/grpc-rs/tree/master/grpc-sys

**Pull request merged**